src/fsharp/FSharp.Core/array.fs

```diff
@@ -957,9 +957,13 @@ namespace Microsoft.FSharp.Collections
             accv
 
         [<CompiledName("Average")>]
-        let inline average      (array:_[]) = 
+        let inline average      (array:'T[]) = 
             checkNonNull "array" array
-            Seq.average array
+            if array.Length = 0 then invalidArg "array" LanguagePrimitives.ErrorStrings.InputArrayEmptyString
```

Comment:
This one looks beneficial, I would expect it easy to show good perf results for this.